clash-lib/src/Clash/Driver.hs

```diff
@@ -682,55 +686,49 @@ normalizeEntity
   -- ^ Hardcoded 'Type' -> 'HWType' translator
   -> (PrimStep, PrimUnwind)
   -- ^ Hardcoded evaluator (delta-reduction)
-  -> [Id]
+  -> [TopEntityT]
   -- ^ TopEntities
   -> ClashOpts
   -- ^ Debug information level for the normalization process
   -> Supply.Supply
   -- ^ Unique supply
   -> Id
   -- ^ root of the hierarchy
-  -> BindingMap
-normalizeEntity reprs bindingsMap primMap tcm tupTcm typeTrans eval topEntities
+  -> (BindingMap, VarEnv TopEntityT)
+normalizeEntity reprs bindingsMap primMap tcm tupTcm typeTrans eval topEnts
   opts supply tm = transformedBindings
   where
-    doNorm = do norm <- normalize [tm]
-                let normChecked = checkNonRecursive norm
-                cleaned <- cleanupGraph tm normChecked
-                return cleaned
+    doNorm = do
+      norm <- normalize [tm]
+      let normChecked = checkNonRecursive norm
+      cleaned <- cleanupGraph tm normChecked
+      topEnts1 <- Lens.use topEntities
+      return (cleaned, topEnts1)
+
     transformedBindings = runNormalization opts supply bindingsMap
                             typeTrans reprs tcm tupTcm eval primMap emptyVarEnv
-                            topEntities doNorm
+                            topEnts doNorm
 
 -- | topologically sort the top entities
 sortTop
   :: BindingMap
-  -> [( Id
-      , Maybe TopEntity
-      , Maybe Id
-      )]
-  -- ^ topEntity bndr
-  -- + (maybe) TopEntity annotation
-  -- + (maybe) testBench bndr
-  -> [( Id
-      , Maybe TopEntity
-      , Maybe Id
-      )]
-  -- ^ topEntity bndr
-  -- + (maybe) TopEntity annotation
-  -- + (maybe) testBench bndr
-sortTop bindingsMap topEntities =
-  let (nodes,edges) = unzip (map go topEntities)
+  -> [TopEntityT]
+  -> [TopEntityT]
+sortTop bindingsMap topEnts =
+  let (nodes,edges) = unzip (map go topEnts)
   in  case reverseTopSort nodes (concat edges) of
         Left msg   -> error msg
         Right tops -> tops
  where
-  go t@(topE,_,tbM) =
+  go t@(TopEntityT topE _ tbM) =
     let topRefs = goRefs topE topE
         tbRefs  = maybe [] (goRefs topE) tbM
     in  ((varUniq topE,t)
-         ,map ((\(v,_,_) -> (varUniq topE, varUniq v))) (tbRefs ++ topRefs))
+         ,map ((\top -> (varUniq topE, varUniq (topId top)))) (tbRefs ++ topRefs))
 
   goRefs top i =
     let cg = callGraph bindingsMap i
-    in  filter (\(v,_,_) -> v /= top && v /= i && v `elemVarEnv` cg) topEntities
+    in
+      filter
+        (\(TopEntityT v _ _) -> v /= top && v /= i && v `elemVarEnv` cg)
```

Comment:
Even though it's more duplication, I also think using the `topId` field name is clearer here. Every time I see something like this I always get sidelined looking up the data type to see what's actually being used.